components/support/rc_crypto/nss/src/util.rs

```diff
@@ -3,184 +3,165 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use crate::error::*;
-use crate::ErrorKind::NSSUninitialized;
 use nss_sys::*;
-use std::{ffi::CString, os::raw::c_char, sync::Once};
+use std::{ffi::CString, os::raw::c_char};
+
+use std::path::PathBuf;
+use std::sync::OnceLock;
 
 #[cfg(feature = "keydb")]
 use crate::pk11::slot;
-#[cfg(feature = "keydb")]
-use once_cell::sync::OnceCell;
-#[cfg(feature = "keydb")]
-use std::{fs, path::Path};
 
 // This is the NSS version that this crate is claiming to be compatible with.
 // We check it at runtime using `NSS_VersionCheck`.
 pub const COMPATIBLE_NSS_VERSION: &str = "3.26";
 
-static NSS_INIT: Once = Once::new();
-#[cfg(feature = "keydb")]
-static NSS_PROFILE_PATH: OnceCell<String> = OnceCell::new();
-
-// Check whether either thr Rust-internal NSS initialization has run, or NSS has been initialized
-// in the outside.
-#[cfg(not(feature = "keydb"))]
-fn is_nss_initialized() -> bool {
-    NSS_INIT.is_completed() || unsafe { NSS_IsInitialized() == PR_TRUE }
+pub fn expect_nss_initialized() -> Result<()> {
+    INITIALIZED
+        .get()
+        .map(|_| {})
+        .ok_or(ErrorKind::NSSUninitialized.into())
 }
 
-// Check whether either thr Rust-internal NSS initialization has run, or NSS has been initialized
-// in the outside.
-// nss can be initialized either by `ensure_nss_initialized` or by
-// `ensure_initialized_with_profile_dir` (by using the `keydb` feature)
-#[cfg(feature = "keydb")]
-fn is_nss_initialized() -> bool {
-    NSS_INIT.is_completed()
-        || NSS_PROFILE_PATH.get().is_some()
-        || unsafe { NSS_IsInitialized() == PR_TRUE }
+// This and many other nss init code were either taken directly from or are inspired by
+// https://github.com/mozilla/neqo/blob/b931a289eee7d62c0815535f01cfa34c5a929f9d/neqo-crypto/src/lib.rs#L73-L77
+enum NssLoaded {
+    External,
+    NoDb,
+    #[cfg(feature = "keydb")]
+    Db,
 }
 
-pub fn expect_nss_initialized() -> Result<()> {
-    if is_nss_initialized() {
-        Ok(())
-    } else {
-        Err(NSSUninitialized.into())
+static INITIALIZED: OnceLock<Result<NssLoaded>> = OnceLock::new();
```

Comment:
I kind of think this should be `OnceLock<NssLoaded>`, without the `Result`.

This means `init_once` would just return `NssLoaded` and panic on failure, like the previous code did.  There's nothing we can realistically do to recover from that.  Also, it means less noise in the PR, since you don't need to change all the call sites.

Reply:
yes!

I'd say that then also `expect_nss_initialized` should become `assert_nss_initialized` and panic the same way, wdyt?

Reply:
I think `expect_foo` implies a panic on failure - ie, `expect_foo` and `assert_foo` seem almost synonyms?

Reply:
ok I see (already renamed ¯\_(ツ)_/¯)